src/CoreGraphics/CGEvent.cs

```diff
@@ -34,23 +34,70 @@ namespace CoreGraphics {
 
+#if NET
+		class TapData {
+			public TapData (CGEventTapCallback cb, IntPtr userInfo)
+			{
+				Callback = cb;
+				UserInfo = userInfo;
+			}
+
+			public CGEventTapCallback Callback { get; private set; }
+			public IntPtr UserInfo { get; private set; }
+		}
+
+		[UnmanagedCallersOnly]
+		static IntPtr TapCallback (IntPtr tapProxyEvent, CGEventType eventType, IntPtr eventRef, IntPtr userInfo)
+		{
+			var gch = GCHandle.FromIntPtr (userInfo);
+			var tapData = (TapData)gch.Target!;
+			return tapData.Callback (tapProxyEvent, eventType, eventRef, tapData.UserInfo);
+		}
+		
+		[DllImport (Constants.ApplicationServicesCoreGraphicsLibrary)]
+		extern static unsafe IntPtr CGEventTapCreate (CGEventTapLocation location, CGEventTapPlacement place, CGEventTapOptions options, CGEventMask mask, delegate* unmanaged<IntPtr, CGEventType, IntPtr, IntPtr, IntPtr> cback, IntPtr data);
+#else
 		[DllImport (Constants.ApplicationServicesCoreGraphicsLibrary)]
 		extern static IntPtr CGEventTapCreate (CGEventTapLocation location, CGEventTapPlacement place, CGEventTapOptions options, CGEventMask mask, CGEventTapCallback cback, IntPtr data);
+#endif
 
 		public static CFMachPort? CreateTap (CGEventTapLocation location, CGEventTapPlacement place, CGEventTapOptions options, CGEventMask mask, CGEventTapCallback cback, IntPtr data)
 		{
+#if NET
+			IntPtr r;
+			unsafe {
+				var tapData = new TapData (cback, data);
+				var gch = GCHandle.Alloc (tapData);
+				r = CGEventTapCreate (location, place, options, mask, &TapCallback, GCHandle.ToIntPtr (gch));
+			}
+#else
 			var r = CGEventTapCreate (location, place, options, mask, cback, data);
+#endif
 			if (r == IntPtr.Zero)
 				return null;
 			return new CFMachPort (r, true);
 		}
 
+#if NET
+		[DllImport (Constants.ApplicationServicesCoreGraphicsLibrary)]
+		extern static unsafe IntPtr CGEventTapCreateForPSN (IntPtr processSerialNumer, CGEventTapLocation location, CGEventTapPlacement place, CGEventTapOptions options, CGEventMask mask, delegate* unmanaged<IntPtr, CGEventType, IntPtr, IntPtr, IntPtr> cback, IntPtr data);
+#else
 		[DllImport (Constants.ApplicationServicesCoreGraphicsLibrary)]
 		extern static IntPtr CGEventTapCreateForPSN (IntPtr processSerialNumer, CGEventTapLocation location, CGEventTapPlacement place, CGEventTapOptions options, CGEventMask mask, CGEventTapCallback cback, IntPtr data);
+#endif
 		
 		public static CFMachPort? CreateTap (IntPtr processSerialNumber, CGEventTapLocation location, CGEventTapPlacement place, CGEventTapOptions options, CGEventMask mask, CGEventTapCallback cback, IntPtr data)
 		{
+#if NET
+			IntPtr r;
+			unsafe {
+				var tapData = new TapData (cback, data);
+				var gch = GCHandle.Alloc (tapData);
+				r = CGEventTapCreateForPSN (processSerialNumber, location, place, options, mask, &TapCallback, GCHandle.ToIntPtr (gch));
+			}
+#else
 			var r = CGEventTapCreateForPSN (processSerialNumber, location, place, options, mask, cback, data);
```

Comment:
There doesn't seem to be any tests for GCEvent.CreateTap, so it would be good to have that to ensure the new (and old for that matter) code is working as expected.

Reply:
I'll add those in a separate PR.